### PR TITLE
Add promises

### PR DESCRIPTION
**Pull request by hayd**
Move deferred from util/ to promises/.

Add delay and timeout promises.

cc @ry @keroxp #188 (also https://github.com/denoland/deno/issues/1800 https://github.com/denoland/deno_std/issues/201 etc.)

Note: This would also need promises to be added to the registry: https://github.com/denoland/registry/pull/57

_I am sure there are several other useful promise constructs, but these were the first that came to mind._

### CODE REVIEW

File: promises/test.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      promise: null,
+      resolve: () => {},
+      reject: () => {}
+    }) === false
+  );
+});
+
+test(async function delayed() {
+  const t = await timeTaken(async () => await delay(200));
+  assert(t > 100);
```
Comment (bartlomieju) [inline promises/test.ts]: just a nit-pick, isn't it guaranteed that timeout will take at least 200 ms?

Comment (hayd) [reply promises/test.ts]: Should be! I was just being cautious here... I think it's _probably_ okay to decrease 😬 

All these tests pass if I use >= 200 (on each of them).

**Comment by ry** [comment]
What else is going to go into promises?

Can you explain why you're against "util"?

**Comment by hayd** [comment]
Perhaps a pool implementation (which I have) and likely some other helpers (see e.g. [bluebird](http://bluebirdjs.com/docs/api-reference.html)).

util is not descriptive... at all. IMOT util is a blight (everything's a util 😭 ), naming is really important (and difficult), likely we'll refactor in the future but nevertheless we should make a best effort.

Helpers for promises, as well as for generators (ala itertools), are things that should be in std. I remember there was an issue somewhere to discuss what would be in std - but I can't find it.

**Comment by keroxp** [comment]
I have neutral opinion about `util`. But I don't strongly oppose `util` module set. 
A problem we have to decide at now is how should small modules  be placed on.  Those modules are usually difficult to be classified.

There are some different solutions:
- 1. treating every single small module as a top level module set (like //deferred/mod.ts)
- 2. create `util` module set and move them into `util` (//util/deferred.ts)
- 3. create new module sets for them and strictly classify it (//promises/deferred.ts)

I don't agree with 1, and 3 may become difficult to manage as time goes by.
I'm sure 2 is not best solution but possible one at this time.
There are no plan to use `promises` or `promise` for other promise-related modules. 

I respect @ry 's decision for this repository. We have to make some decision on #130 together.

And by this case, I'm also concerned about usage of deno_std's namespace in  deno.land/x/. I will open another issue...

**Comment by hayd** [comment]
An argument could be made for dropping all the independent std modules in the registry, i.e. reference std solely under /x/std/... I'm not sure about that. (Though it would solve the relative/not found issue.)

This PR is independent of #130. There's certainly more promise-related helpers that could be included to /promises/ (I added a couple already and have some follow up ideas), I don't think anything is gained by having deferred in its own file in this case.

There is a larger question/issue: what else should be in std?

**Comment by ry** [comment]
I think we ought to remove Deno.land/x/http and other sub modules of std... I will do some work on this today. 
I very much agree with @keroxp’s position in these org discussions.

File: promises/mod.ts [hunk 2]
```diff
@@ -32,11 +32,26 @@ export function defer<T>(): Deferred<T> {
...
+    setTimeout(() => {
+      resolve(value);
+    }, ms);
+  });
+}
+
+export class TimeoutError extends Error {
+  constructor(ms: number) {
+    const message = `Timeout after ${ms}ms`;
+    super(message);
```
Comment (j-f1) [inline promises/mod.ts]: Could these two lines be combined?

```suggestion
    super(`Timeout after ${ms}ms`);
```

Also, should this use something like Node’s [`ms`](https://npm.im/ms) utility to display the elapsed time in a more friendly manner?

**Comment by hayd** [comment]
@ry should this also be in individual files: promises/deferred.ts promises/timeout.ts promises/delay.ts ??

**Comment by ry** [comment]
This one is out of date - if you want to land this still - let's discuss on gitter

**Comment by hayd** [comment]
@ry defer is useful (and already used in several projects), as are the others...

Happy to re-land this, but want to know how you want it structured.

>  should this also be in individual files: promises/deferred.ts promises/timeout.ts promises/delay.ts

**Pull request closed without merging**